server/src/main/java/org/elasticsearch/index/IndexMode.java

```diff
@@ -74,6 +81,28 @@ void validateWithOtherSettings(Map<Setting<?>, Object> settings) {
             }
         }
 
+        @Override
+        public void validateTimestampRange(DocumentParserContext context, long timestamp) {
+            if (context.mappingLookup().getMapper(TIMESTAMP_FIELD).typeName().equals(DateFieldMapper.DATE_NANOS_CONTENT_TYPE)) {
+                timestamp /= NSEC_PER_MSEC;
+            }
+
+            long startTime = context.indexSettings().getTimeSeriesStartTime();
+            long endTime = context.indexSettings().getTimeSeriesEndTime();
+
+            if (timestamp < startTime) {
+                throw new IllegalArgumentException(
+                    "time series index " + TIMESTAMP_FIELD + " value [" + timestamp + "] must be larger than " + startTime
+                );
+            }
+
+            if (timestamp >= endTime) {
+                throw new IllegalArgumentException(
+                    "time series index " + TIMESTAMP_FIELD + " value [" + timestamp + "] must be smaller than " + endTime
+                );
```

Comment:
You pinged me and asked for another review but haven't change this so I'll restate my objections - I feel like, as much as possible, `IndexMode` should be about making sure that the appropriate settings are set and the appropriate mapping is in place. Then we can let the mapping do the assertions.  So I think it's more right to do the test in `DataStreamTimestampFieldMapper` and do it something like:
```
if (context.settings.startTime.isSet) {
  if (time < context.settings.startTime) {
...
```

I like this because a reader doesn't have to know what time series mode is - they just have to know that they is some way to set the min and max time. If they need to figure out how to set it then they can. But lots of times I find myself not caring where a behavior comes from, just that we have this behavior.